### PR TITLE
Fix #2198: Introduce pre push hook for ktlint

### LINKED ISSUE #2198: Introduce pre push hook for Kotlin
**Is your feature request related to a problem? Please describe.**
We are linting our kotlin files using ktlint and there are quite a number of times when we have ktlint failure in GitHub checks. 

**Describe the solution you'd like**
pre-push will prevent all such failures.

### PR DESCRIPTION
**Pull request by anandwana001**
## Explanation

Fix #2198 

As we know that we have ktlint check in our actions and we also enforce for good formatting code. So, rather than sending a PR and then fixing the ktlint issue, we are introducing a pre-push hook which will run the ktlint before any push. 
As per this PR, we want any contributor after opening the codebase we want them to run a script which will copy our hooks script into the `.git/hooks` folder. 

Script to run -> `bash scripts/setup.sh`

## Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

### CODE REVIEW

**Review by rt4914 — APPROVED** [review]
Screenshot 2020-10-13 at 12 06 30 PM


Tested this and it works correctly. Thanks @anandwana001 

Will need to defer to @BenHenning and @vinitamurthi for code correctness.

**Comment by BenHenning** [comment]
I'll need to take a look at this tomorrow my time (it's getting a bit late). @vinitamurthi it might be nice if you could take a look at this first. One issue with the pre-commit hook is that we can't force people to actually install it in the Android workflow, and if you don't install it it's not as useful. :)

File: scripts/pre-commit
Comment (vinitamurthi) [inline scripts/pre-commit]: Nit: Can we order them alphabetically i.e. app/data/domain/testing/utility

File: scripts/setup.py
Comment (vinitamurthi) [inline scripts/setup.py]: The downside to this is that, if we make any changes to the pre-commit hook, it will not get updated to folks who already have the hook installed. Maybe we should not do this and just replace the file even if it exists (Since this script shouldn't be run very often anyway)

**Comment by vinitamurthi** [comment]
> One issue with the pre-commit hook is that we can't force people to actually install it in the Android workflow, and if you don't install it it's not as useful. :)

That's actually why I thought we should install the hook in a setup script (among other reasons). Today the setup script is not something we can force people to do, but in the future we can actually have people use the setup script to install everything they require to get oppia-android working (particularly, bazel :) ) 
Today since we only need Android Studio which has gradle along with it, the setup script is an 'extra' script that is not really important. But when we move to Bazel, we can actually have the setup script do the following:
1) Install Bazel (which makes it essential)
2) Install Ktlint
3) Install Buf
4) Install whatever XML/java linters we chose to use
5) Install codecoverage libraries if need be
6) Add the pre-commit hook

And in documentation we basically have the installation steps as: Clone the repo, download Android studio, run this setup script.

File: scripts/pre-commit
Comment (BenHenning) [inline scripts/pre-commit]: Maybe output the suggested auto-fix line to make this easier for contributors. I personally always use ``-F`` and only fix the things ktlint can't.

File: scripts/setup.py
Comment (BenHenning) [inline scripts/setup.py]: One concern I have is this now requires Python to be installed. Our current process is Python-independent until now (consider most Android developers won't be running the backend server, and many won't be using Ubuntu that often comes with Python preinstalled).

Two alternative options
1. Use Kotlin with Bazel & bash (bash would just run a Bazel command to run the script where the script is written in Kotli)
2. Use native Bash

The former may be more work, but's probably more future-proof as the setup script gets more complex.

@vinitamurthi @rt4914 any thoughts?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @anandwana001! Left some high-level thoughts.

File: scripts/setup.py
Comment (anandwana001) [reply scripts/setup.py]: With native bash, we can have a simple script of maybe 2 line for moving/copying of pre commit file.

File: scripts/pre-commit
Comment (anandwana001) [reply scripts/pre-commit]: This will be helpful. I will add it.

File: scripts/setup.py
Comment (rt4914) [reply scripts/setup.py]: I think Bash can be a good option. As we have more things like this we can integrate that in bash file.

File: scripts/pre-commit
Comment (anandwana001) [reply scripts/pre-commit]: Done

File: scripts/setup.py
Comment (anandwana001) [reply scripts/setup.py]: now with bash, what we are doing is moving of file.

File: scripts/pre-commit
Comment (anandwana001) [reply scripts/pre-commit]: Done

File: scripts/setup.py
Comment (anandwana001) [reply scripts/setup.py]: Done

File: scripts/pre-commit
Comment (anandwana001) [inline scripts/pre-commit]: @BenHenning do you want me to add a comment here, what does the `-F` is doing in particular?

File: scripts/pre-commit
Comment (vinitamurthi) [inline scripts/pre-commit]: I'm being a bit nitpicky (sorry!) - can we have the entire lint check work within a function and then call that function in this script? The reason I ask for that is so that we can cleanly extend this hook if we want to add more checks into it later on

**Review by vinitamurthi — CHANGES_REQUESTED** [review]

File: scripts/pre-commit
Comment (BenHenning) [reply scripts/pre-commit]: Hmm. This actually wasn't what I had in mind. I was thinking that the pre-commit hook would *fail* if there are linter issues, and during failure would print out this line (with the ``-F``) so that the user could use it to fix most of the issues. I don't think we should auto-fix linter issues (this falls under the general idea that we should never automatically change people's code without their explicit request).

@vinitamurthi @rt4914 any thoughts?

File: scripts/pre-commit
Comment (BenHenning) [reply scripts/pre-commit]: One step further: I think the lint check itself should be its own script that this hook calls (apologies if that's what you meant @vinitamurthi). The pre-commit hook ought to just be a set of commands to run, and having the command separate also lets people use it directly.

**Review by BenHenning — COMMENTED** [review]
Thanks @anandwana001! Left a few more thoughts.

File: scripts/pre-commit
Comment (vinitamurthi) [reply scripts/pre-commit]: I agree with not modifying without the user's explicit approval. So we should run ktlint without -F , mention the errors, and then mention how to fix them (i.e. use the -F flag  as Ben mentioned)

File: scripts/pre-commit
Comment (vinitamurthi) [reply scripts/pre-commit]: Yes that's what I meant sorry I didn't make it clear! We should keep them in separate scripts and the pre-commit hook should have commands to run those scripts one after the other

File: scripts/pre-commit
Comment (rt4914) [reply scripts/pre-commit]: I agree that we should not change code automatically as personally I think that can lead to bad experience especially new comers.

File: scripts/pre-commit
Comment (anandwana001) [reply scripts/pre-commit]: Sounds good, Thanks @vinitamurthi

File: scripts/ktlint-pre-commit
Comment (vinitamurthi) [inline scripts/ktlint-pre-commit]: Lets remove the -F flag here

File: scripts/ktlint-pre-commit
Comment (vinitamurthi) [inline scripts/ktlint-pre-commit]: Change the above text to - "Please note, there mChight be a possibility where the above commend will not fix the issue. In that case, you will have to fix it yourself.

File: scripts/ktlint-pre-commit
Comment (vinitamurthi) [inline scripts/ktlint-pre-commit]: Instead of calling this ktlint-pre-commit, let's maybe call it ktlint-check ? or kotlin-lint-check? Want to remove the 'pre-commit' part out of it

File: scripts/pre-commit
Comment (vinitamurthi) [inline scripts/pre-commit]: Sorry I just noticed this -- should we be doing it as a pre-commit check? I think pre-push may be a bit better @BenHenning any thoughts?

**Review by vinitamurthi — CHANGES_REQUESTED** [review]

File: scripts/ktlint-pre-commit
Comment (anandwana001) [reply scripts/ktlint-pre-commit]: Done, resolving on behalf of Vinita as it's a small fix.

File: scripts/ktlint-pre-commit
Comment (anandwana001) [reply scripts/ktlint-pre-commit]: Done, resolving on behalf of Vinita as it's a small fix.

File: scripts/ktlint-pre-commit
Comment (anandwana001) [reply scripts/ktlint-pre-commit]: Done, resolving on behalf of Vinita as it's a small fix.

File: scripts/pre-commit
Comment (BenHenning) [reply scripts/pre-commit]: Ah yes, I agree. Pre-commit is a bit too often I think--let's do pre-push instead.

**Comment by BenHenning** [comment]
I left a reply to @vinitamurthi's question and will follow up with a full review once this is ready to be reviewed again.

File: scripts/pre-commit
Comment (anandwana001) [reply scripts/pre-commit]: Done

File: scripts/pre-commit
Comment (anandwana001) [reply scripts/pre-commit]: Done

**Review by vinitamurthi — APPROVED** [review]
LGTM, thanks @anandwana001 !

File: scripts/pre-push
Comment (BenHenning) [inline scripts/pre-push]: TODOs should be of the format ``TODO(#1234): Actionable item.``

Please fix. Each TODO should have an issue associated with it.

File: scripts/pre-push
Comment (BenHenning) [inline scripts/pre-push]: Does this actually work? It seems like in the setup we only copy the pre-push, not the lint check script.

File: scripts/kotlin-lint-check
Comment (BenHenning) [inline scripts/kotlin-lint-check]: This file should end with .sh since it's a shell file.

File: scripts/kotlin-lint-check
Comment (BenHenning) [inline scripts/kotlin-lint-check]:
```suggestion
  echo "Please fix the above issues. You can also use the ktlint -F --android path/to/file.kt
  command to fix the most common issues."
```

File: scripts/kotlin-lint-check
Comment (BenHenning) [inline scripts/kotlin-lint-check]: Please update the suggested ``ktlint`` command to just fix all issues in the codebase (e.g. by passing all of the module folders to it).

File: scripts/pre-push
Comment (BenHenning) [inline scripts/pre-push]: Comment seems unnecessary since the line is fairly self-explanatory.

File: scripts/pre-push
Comment (BenHenning) [inline scripts/pre-push]: Is this meant to be a list of items? If so, prefer making this a markdown list, e.g.

```bash
# This script will run the pre-push checks in the following order:
#  - ktlint
#  - (others in the future)
```

File: scripts/pre-push
Comment (BenHenning) [inline scripts/pre-push]: These don't seem like instructions? Suggest removing this.

File: scripts/setup.sh
Comment (BenHenning) [inline scripts/setup.sh]: Do we also need to make the hook executable?

File: scripts/setup.sh
Comment (BenHenning) [inline scripts/setup.sh:11]: I think one of my earlier concerns is still a problem: how do we actually get people to do this if everyone is using Android Studio by default?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @anandwana001! Did a full review.

File: scripts/pre-push
Comment (anandwana001) [reply scripts/pre-push]: Why do we need to copy the lint check script into the hooks folder?
the flow is like:
1. setup will be exec by the contributor
2. copy the pre-push to hooks
3. pre-push file will run the scripts in the scripts folder

As per our discussion in the above comments, the scripts cane be executable externally also.

File: scripts/setup.sh
Comment (anandwana001) [reply scripts/setup.sh:11]: This is something anyone has to do it by themself after initial codebase setup. 

reference - https://github.com/oppia/oppia-android/pull/2002#issuecomment-707601602

File: scripts/pre-push
Comment (BenHenning) [reply scripts/pre-push]: Sorry, to be clear I don't think we should copy the scripts into the hooks folder. I'm actually asking if this setup works because the hook doesn't correctly reference the scripts folder. It's assuming that the kotlin-lint-check script is either in the same directly or somewhere else on the path. Am I missing something?

File: scripts/setup.sh
Comment (BenHenning) [reply scripts/setup.sh:11]: Okay, in that case I think this is fine. Vinita's plan is a good one, and addresses my concern. Thanks!

**Comment by rt4914** [comment]
Unassigned myself because of correctness of code I will have to defer to @BenHenning and @vinitamurthi

File: scripts/pre-push
Comment (anandwana001) [reply scripts/pre-push]: done

File: scripts/kotlin-lint-check
Comment (anandwana001) [reply scripts/kotlin-lint-check]: done

File: scripts/kotlin-lint-check
Comment (anandwana001) [reply scripts/kotlin-lint-check]: done

File: scripts/kotlin-lint-check
Comment (anandwana001) [reply scripts/kotlin-lint-check]: done

File: scripts/pre-push
Comment (anandwana001) [reply scripts/pre-push]: done

File: scripts/pre-push
Comment (anandwana001) [reply scripts/pre-push]: done

File: scripts/pre-push
Comment (anandwana001) [reply scripts/pre-push]: done

File: scripts/setup.sh
Comment (anandwana001) [reply scripts/setup.sh]: yes, the hook (pre-push file) is executable.

File: scripts/pre-push
Comment (anandwana001) [reply scripts/pre-push]: ops, missed it. Done now.

**Comment by rt4914** [comment]
@anandwana001 We discussed that if possible I can resolve ben's comment but even if I do that we will need approval from @BenHenning and therefore I am not resolving his comments. Thanks.

File: scripts/setup.sh
Comment (BenHenning) [reply scripts/setup.sh]: Ah, but will Git retain the +x bit for files downloaded by other users? I think in Oppia web we use chmod +x to ensure the file is executable. Can you make sure we're following the same strategy for these files?

Or, do we not need the +x bit for shell files or Git hooks?

**Comment by BenHenning** [comment]
Thanks @anandwana001. Just had one follow-up regarding the executable bit of the pre-commit file, otherwise this looks good to me.

File: scripts/setup.sh
Comment (anandwana001) [reply scripts/setup.sh]: No, I don't think we need chmod, I am copying the file in such a way where it becomes executable.

Comment (BenHenning) [reply scripts/setup.sh]: Okay, let's check it in and see then. Thanks :)

File: .idea/gradle.xml
Comment (BenHenning) [inline .idea/gradle.xml]: Please don't include unrelated config changes here.

Also, reminder to make sure file changes include EOFs (below).

File: scripts/kotlin-lint-check.sh
Comment (BenHenning) [inline scripts/kotlin-lint-check.sh]: I realized this quite late, but this requires ktlint to be in the user's path (which may not always be the case for everyone).

I think we should actually set up ktlint for users as part of the setup script. It seems like we can make it work for both Unix & Windows by executing ktlint as a Jar file directly (e.g. 'java -jar') per its documentation. We just need to download it to a utility outside of the oppia-android repository.

On web, we suggest people put the oppia repo in a folder called 'opensource'. I think we should continue that pattern here, that way we can introduce a new directory structure that simplifies this:
- opensource/oppia-android
- opensource/oppia-android-tools (this isn't a repo and is where things like ktlint will go)

What do you think?

**Review by BenHenning — CHANGES_REQUESTED** [review]
Thanks @anandwana001! I was about to approve then realized one last thing. :( We should make setup.sh the one-stop-solves-all solution for preparing the codebase. To start, it should set up ktlint and in the future we can add other tools (such as Bazel).

File: .idea/gradle.xml
Comment (anandwana001) [reply .idea/gradle.xml]: ah, committed by mistake as update the AS.

File: scripts/kotlin-lint-check.sh
Comment (anandwana001) [reply scripts/kotlin-lint-check.sh]: Sounds good, I am also using the same pattern 
Screenshot 2020-12-10 at 15 42 11

Comment (anandwana001) [reply scripts/kotlin-lint-check.sh]: I am thinking to copy start.py from Oppia so that I will get a conditional statement like `common.is_linux_os()` `elif common.is_mac_os()` and else `windows` 
So, the flow will be -> download release file, and run the commands with jar fil. 
Windows -> `java -jar ktlint`
Linux -> `java -jar ktlint`
and for mac, I am thinking to do `brew install ktlint` or should we go with the jar file option only and run commands like `java -jar ktlint`?

https://github.com/oppia/oppia/blob/0068ec7614c31a9389739379943ccdfba0f7f6e0/scripts/start.py#L174

Comment (FareesHussain) [inline scripts/kotlin-lint-check.sh:7]: I've been exploring this pre push hook for ktlint 

if we run `./gradlew ktlintcheck` instead of ktlint it provides references to the particular line of code where we need to change 
wouldn't this be better
image

hoping this  doesn't have any issues with bazel

Comment (FareesHussain) [inline scripts/kotlin-lint-check.sh]: I've been exploring this pre push hook for ktlint 

if we run `./gradlew ktlintcheck` instead of ktlint it provides references to the particular line of code where we need to change 
wouldn't this be better
image

hoping this  doesn't have any issues with bazel

**Review by FareesHussain — COMMENTED** [review]
Small doubt regarding how to run ktlint in pre push hook

File: scripts/kotlin-lint-check.sh
Comment (FareesHussain) [reply scripts/kotlin-lint-check.sh]: Ktlint Gradle - https://github.com/jlleitschuh/ktlint-gradle

I hope using this doesn't require the user to setup ktlint seperately

Comment (anandwana001) [reply scripts/kotlin-lint-check.sh]: We are on the migration from gradle to bazel build system, so we will be removing gradle soon, that's the reason we are keeping it gradle free.

File: scripts/setup.sh
Comment (FareesHussain) [inline scripts/setup.sh:9]: I've run this script on windows and it works

Comment (anandwana001) [reply scripts/setup.sh:9]: Thanks for checking.

Comment (anandwana001) [reply scripts/setup.sh:9]: Will document that windows user need git bash command line to run this script.

File: scripts/kotlin-lint-check.sh
Comment (anandwana001) [reply scripts/kotlin-lint-check.sh]: Done, using ktlint executable file.

Comment (anandwana001) [reply scripts/kotlin-lint-check.sh]: Screenshot 2020-12-18 at 16 08 21

File: scripts/setup.sh
Comment (anandwana001) [inline scripts/setup.sh]: @BenHenning Let me know if there is any short hand for this

```console
cd ..
mkdir -p oppia-android-tools
cd oppia-android-tools
```

Comment (BenHenning) [reply scripts/setup.sh]: Best I can come up with is:

```bash
mkdir -p ../oppia-android-tools
cd ../oppia-android-tools
```

(this might be slightly better since it's keeping the pwd within a directory that we manage)

**Review by BenHenning — APPROVED** [review]
Thanks @anandwana001! LGTM!

File: scripts/kotlin-lint-check.sh
Comment (BenHenning) [inline scripts/kotlin-lint-check.sh]: Actually, it's a bit weird to be passing ktlint to bash like this. We should be following the ktlint team's instructions here: 

https://github.com/pinterest/ktlint#usage.

The universal command would be:

```sh
java -jar ../oppia-android-tools/ktlint --android app/src/**/*.kt data/src/**/*.kt domain/src/**/*.kt testing/src/**/*.kt utility/src/**/*.kt
```

**Comment by BenHenning** [comment]
@anandwana001 just 2 comments to address from me before merging. PTAL.

File: scripts/setup.sh
Comment (FareesHussain) [inline scripts/setup.sh]: @anandwana001 
As there is issue with running ktlint in windows
can we switch ktlint version to `0.37.1` for prepush hook 
as in the `0.37.1` release they fixed the windows issue - https://github.com/pinterest/ktlint/issues/761

Reference - https://github.com/pinterest/ktlint/releases/tag/0.37.1

Comment (anandwana001) [reply scripts/setup.sh]: (@Srikanth-Kb)

https://github.com/oppia/oppia-android/issues/2352

Comment (srikanthkb) [reply scripts/setup.sh]: @FareesHussain and @anandwana001 Thanks for the suggestions.
I will work on this.

File: scripts/kotlin-lint-check.sh
Comment (anandwana001) [reply scripts/kotlin-lint-check.sh]: Done

File: scripts/setup.sh
Comment (anandwana001) [reply scripts/setup.sh:9]: done

File: scripts/setup.sh
Comment (anandwana001) [reply scripts/setup.sh]: done
resolving on behalf of Ben

File: scripts/kotlin-lint-check.sh
Comment (anandwana001) [reply scripts/kotlin-lint-check.sh]: Resolving on behalf of ben

**Pull request merged**
